libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,291 @@
+    pub fn create(&mut self) -> Result<Self> {
+        let err = unsafe { libbpf_sys::bpf_tc_hook_create(&mut self.hook as *mut _) };
+        // the hook may already exist, this is not an error
+        if err != 0 && err != -(EEXIST as i32) {
+            Err(Error::System(err))
+        } else {
+            Ok(*self)
+        }
+    }
+
+    /// Set the interface to be used
+    ///
+    /// Interfaces can be listed by using ip link command from the iproute2 software package
+    pub fn ifindex(&mut self, idx: i32) -> &mut Self {
+        self.hook.ifindex = idx;
+        self
+    }
+
+    /// Set what type of TC point to attach onto
+    ///
+    /// TC_EGRESS, TC_INGRESS, or TC_CUSTOM
+    ///
+    /// An TC_EGRESS|TC_INGRESS hook can be used as an attach_point for calling
+    /// the destroy() method to remove the clsact bpf tc qdisc, but cannot
+    /// be used to attach()
+    pub fn attach_point(&mut self, ap: TcAttachPoint) -> &mut Self {
+        self.hook.attach_point = ap;
+        self
+    }
+
+    /// Set the parent of a hook
+    /// Will return EINVAL if set upon an TC_EGRESS/TC_INGRESS/TC_EGRESS|TC_INGRESS hook
+    ///
+    /// Must be set on a TC_CUSTOM hook
+    /// Current acceptable values are TC_H_CLSACT for maj, and
+    /// TC_H_MIN_EGRESS or TC_H_MIN_INGRESS for min
+    pub fn parent(&mut self, maj: u32, min: u32) -> &mut Self {
+        /* values from libbpf.h BPF_TC_PARENT() */
+        let parent = (maj & TC_H_MAJ_MASK) | (min & TC_H_MIN_MASK);
+        self.hook.parent = parent;
+        self
+    }
+
+    /// Set whether this hook should replace an existing hook
+    ///
+    /// If replace is not true upon attach, and a hook already exists
+    /// an EEXIST error will be returned from attach()
+    pub fn replace(&mut self, replace: bool) -> &mut Self {
+        if replace {
+            self.opts.flags = BPF_TC_F_REPLACE;
+        } else {
+            self.opts.flags = 0;
+        }
+        self
+    }
+
+    /// Set the handle of a hook.
+    /// If unset upon attach, the kernel will assign a handle for the hook
+    pub fn handle(&mut self, handle: u32) -> &mut Self {
+        self.opts.handle = handle;
+        self
+    }
+
+    /// Set the priority of a hook
+    /// if unset upon attach, the kernel will assign a priority for the hook
+    pub fn priority(&mut self, priority: u32) -> &mut Self {
+        self.opts.priority = priority;
+        self
+    }
+
+    /// Query a hook to inspect the program identifier (prog_id)
+    pub fn query(&mut self) -> Result<u32> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_query(&self.hook as *const _, &mut opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(opts.prog_id)
+        }
+    }
+
+    /// Attach the TcHook so that the program starts processing
```

Comment:
Would you mind clarifying this to  "Attach a filter to a TcHook"?

Reply:
Resolved